kernel/sched.c

```diff
@@ -100,13 +100,22 @@ bool z_is_t1_higher_prio_than_t2(struct k_thread *thread_1,
 		int32_t d1 = thread_1->base.prio_deadline;
 		int32_t d2 = thread_2->base.prio_deadline;
 
-		return (d2 - d1) >= 0;
+		return (d2 - d1) > 0;
 	}
 #endif
 
 	return false;
 }
 
+bool z_is_t1_higher_prio_than_t2(struct k_thread *thread_1,
+				 struct k_thread *thread_2)
+{
+	const bool ret = is_t1_higher_prio_than_t2(thread_1, thread_2);
+	/* If a>b and b>a, that's a problem */
+	__ASSERT_NO_MSG(!ret || !is_t1_higher_prio_than_t2(thread_2, thread_1));
```

Comment:
FWIW: I'd suggest not putting an assert in the deepest part of the scheduler loop and instead write a test case that shows up the bug you found.

Reply:
Hm. Ok.

I'll have to keep that in mind - I'm used to a much more assert-heavy codebase than Zephyr is.

- [ ] Remove assert.
- [ ] Add yield test case.

Reply:
It's a judgement call I guess.  We have asserts enabled for basically everything running under tests/, the idea is that it's a routine sort of thing you can run with all the time during development.  And doubling the cost of the scheduler priority predicate seems like it's over the line as far as impact goes.  In this case it seems like a "who ran before who" test should be a simple and welcome addition to the regression suite with equal value and zero overhead.

Reply:
FWIW: I'd be happy to approve the patch without the assert or the test.  But the test would be nice. :)

Reply:
Ah, I get the distinction now.

I'm used to `ASSERT` / `SLOW_ASSERT`, where `SLOW_ASSERT` is exactly what it says on the tin.

You're saying if I understand correctly that you don't have `SLOW_ASSERT` and try to ensure test coverage instead.

Reply:
As an aside, in my experience asserts / slowasserts are often better than tests for this sort of thing, for a combination of several reasons:

1. Adding a test for this case only catches this one specific case. If someone else e.g. adds another tiebreaker to `is_t1_higher_prio_than_t2` and misses an edge case, the testcases won't catch that. The testcase only catches the specific case of "deadline priority tiebreaker not working".
2. There's a lot of "free" test coverage that you get when you have asserts for invariants. Your context-switch stresstest that someone wrote to test something else entirely that happens to break this invariant? Actually fails.
3. The external visibility of an invariant failing somewhere deep in the bowels of the codebase is often slim-to-none. (E.g. if this had been an issue that only occurred across multiple processors... good luck testing it. The current approach only works because it's a single processor in cooperative mode.)